frappe/api/v2.py

```diff
@@ -0,0 +1,193 @@
+	# evaluate frappe.get_list
+	return frappe.call(frappe.client.get_list, doctype, **frappe.form_dict)
+
+
+def count(doctype: str) -> int:
+	from frappe.desk.reportview import get_count
+
+	frappe.form_dict.doctype = doctype
+
+	return get_count()
+
+
+def create_doc(doctype: str):
+	data = frappe.form_dict
+	data.pop("doctype", None)
+	return frappe.new_doc(doctype, **data).insert()
+
+
+def update_doc(doctype: str, name: str):
+	data = frappe.form_dict
+
+	doc = frappe.get_doc(doctype, name, for_update=True)
+	data.pop("flags", None)
+	doc.update(data)
+	doc.save()
+
+	# check for child table doctype
+	if doc.get("parenttype"):
+		frappe.get_doc(doc.parenttype, doc.parent).save()
+
+	return doc
+
+
+def delete_doc(doctype: str, name: str):
+	frappe.client.delete_doc(doctype, name)
+	frappe.response.http_status_code = 202
+	return "ok"
+
+
+def execute_doc_method(doctype: str, name: str, method: str | None = None):
+	"""Get a document from DB and execute method on it.
+
+	Use cases:
+	- Submitting/cancelling document
+	- Triggering some kind of update on a document
+	"""
+	method = method or frappe.form_dict.pop("run_method")
+	doc = frappe.get_doc(doctype, name)
+	doc.is_whitelisted(method)
+
+	doc.check_permission(PERMISSION_MAP[frappe.request.method])
+	return doc.run_method(method, **frappe.form_dict)
+
+
+def run_doc_method(method: str, document: dict[str, Any] | str, kwargs=None):
+	"""run a whitelisted controller method on in-memory document.
+
+
+	This is useful for building clients that don't necessarily encode all the business logic but
+	call server side function on object to validate and modify the doc.
+
+	The doc CAN exists in DB too and can write to DB as well if method is POST.
+	"""
+
+	if isinstance(document, str):
+		document = frappe.parse_json(document)
+
+	if kwargs is None:
+		kwargs = {}
+
+	doc = frappe.get_doc(document)
+	doc._original_modified = doc.modified
+	doc.check_if_latest()
+
+	doc.check_permission(PERMISSION_MAP[frappe.request.method])
+
+	method_obj = getattr(doc, method)
+	fn = getattr(method_obj, "__func__", method_obj)
+	is_whitelisted(fn)
+	is_valid_http_method(fn)
+
+	new_kwargs = get_newargs(fn, kwargs)
+	response = doc.run_method(method, **new_kwargs)
+	frappe.response.docs.append(doc)  # send modified document and result both.
+	return response
+
+
+url_rules = [
+	# RPC calls
+	Rule("/method/login", endpoint=login),
+	Rule("/method/logout", endpoint=logout),
+	Rule("/method/ping", endpoint=frappe.ping),
+	Rule("/method/upload_file", endpoint=upload_file),
+	Rule("/method/<method>", endpoint=handle_rpc_call),
+	Rule(
+		"/method/run_doc_method",
+		methods=["GET", "POST"],
+		endpoint=lambda: frappe.call(run_doc_method, **frappe.form_dict),
+	),
+	Rule("/method/<doctype>/<method>", endpoint=handle_rpc_call),
+	# Document level APIs
+	Rule("/document/<doctype>", methods=["GET"], endpoint=document_list),
+	Rule("/document/<doctype>", methods=["POST"], endpoint=create_doc),
```

Comment:
Doctype list and create should be on `GET /doctype/ToDo` and `POST /doctype/ToDo` respectively?

`/document/ToDo/<name>/*` can be reserved for records that already exist?